nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch.java

```diff
@@ -0,0 +1,326 @@
+        "splitUp-no",
+        "No",
+        "Don't split up results."
+    );
+
+    public static final PropertyDescriptor SPLIT_UP_HITS = new PropertyDescriptor.Builder()
+            .name("el-rest-split-up-hits")
+            .displayName("Split up search results")
+            .description("Split up search results into one flowfile per result.")
+            .allowableValues(SPLIT_UP_HITS_NO, SPLIT_UP_YES)
+            .defaultValue(SPLIT_UP_HITS_NO.getValue())
+            .required(true)
+            .expressionLanguageSupported(false)
+            .build();
+    public static final PropertyDescriptor SPLIT_UP_AGGREGATIONS = new PropertyDescriptor.Builder()
+            .name("el-rest-split-up-aggregations")
+            .displayName("Split up aggregation results")
+            .description("Split up aggregation results into one flowfile per result.")
+            .allowableValues(SPLIT_UP_HITS_NO, SPLIT_UP_YES)
+            .defaultValue(SPLIT_UP_HITS_NO.getValue())
+            .required(true)
+            .expressionLanguageSupported(false)
+            .build();
+
+    public static final PropertyDescriptor CLIENT_SERVICE = new PropertyDescriptor.Builder()
+            .name("el-rest-client-service")
+            .displayName("Client Service")
+            .description("An ElasticSearch client service to use for running queries.")
+            .identifiesControllerService(ElasticSearchClientService.class)
+            .required(true)
+            .build();
+
+    private static final Set<Relationship> relationships;
+    private static final List<PropertyDescriptor> propertyDescriptors;
+
+    private ElasticSearchClientService clientService;
+
+    static {
+        final Set<Relationship> _rels = new HashSet<>();
+        _rels.add(REL_ORIGINAL);
+        _rels.add(REL_FAILURE);
+        _rels.add(REL_HITS);
+        _rels.add(REL_AGGREGATIONS);
+        relationships = Collections.unmodifiableSet(_rels);
+
+        final List<PropertyDescriptor> descriptors = new ArrayList<>();
+        descriptors.add(QUERY);
+        descriptors.add(QUERY_ATTRIBUTE);
+        descriptors.add(INDEX);
+        descriptors.add(TYPE);
+        descriptors.add(CLIENT_SERVICE);
+        descriptors.add(SPLIT_UP_HITS);
+        descriptors.add(SPLIT_UP_AGGREGATIONS);
+
+        propertyDescriptors = Collections.unmodifiableList(descriptors);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return propertyDescriptors;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        clientService = context.getProperty(CLIENT_SERVICE).asControllerService(ElasticSearchClientService.class);
+    }
+
+    @OnUnscheduled
+    public void onUnscheduled() {
+        this.clientService = null;
```

Comment:
This should be changed from "OnUnscheduled"[1] to "OnStopped"[2]. "OnUnscheduled" will be run as soon as the user clicks stop. This stops the processor from being schedueld again but there could still be lingering tasks running (thus leading to null pointer exceptions when it tries to use the clientService). "OnStopped" is run once all tasks finish.

[1] https://github.com/apache/nifi/blob/master/nifi-api/src/main/java/org/apache/nifi/annotation/lifecycle/OnUnscheduled.java
[2] https://github.com/apache/nifi/blob/master/nifi-api/src/main/java/org/apache/nifi/annotation/lifecycle/OnStopped.java